Generate native glibc.modulemap when cross-compiling

karwa opened the pull request:
<!-- Please complete this template before creating pull request. -->
#### What's in this pull request?

<!-- Description about pull request. -->

When cross-compiling, we have to use generate a glibc.modulemap for use at build-time whose headers are in some sysroot folder on our local drives. This generated file obviously won't work once installed.

With this PR, we generate the glibc.modulemap needed for building while at the same time generating a non-prefixed version in `/native-substitutes/glibc.modulemap`. The non-prefixed one is the one which is added to the install list.

I think this issue is also a pain on Android. If somebody could test it, we could add an `or SDK == ANDROID` condition to generate the native modulemap.

<!-- If this pull request resolves any bugs from Swift bug tracker -->

---

<!-- This selection should only be completed by Swift admin -->

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>



The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

<!-- Thank you for your contribution to Swift! -->

springsup committed:
Generate native glibc.modulemap when cross-compiling

jrose-apple commented:
I think the module map should include sysroot-unaware paths; the use of the -sdk option supplies the sysroot.

gribozavr commented on stdlib/public/Platform/CMakeLists.txt:
```diff
@@ -68,11 +69,28 @@ foreach(sdk ${SWIFT_SDKS})
+
+      # If cross-compiling, generate a native modulemap without the sysroot prefix
+      # The prefixed one is still the 'glibc.modulemap' which we'll look at during the build process
+      if(CMAKE_CROSSCOMPILING)
```
I'm afraid this is not the right variable to key off.  CMake does not always know that we are cross-compiling.

Instead, you should be checking the `sdk` and `arch` variables set above.

gribozavr commented on stdlib/public/Platform/CMakeLists.txt:
```diff
@@ -38,23 +38,24 @@ foreach(sdk ${SWIFT_SDKS})
-          #        architecture as part of the path to a distro that does not.
-          set(GLIBC_ARCH_INCLUDE_PATH "${GLIBC_INCLUDE_PATH}/${CMAKE_LIBRARY_ARCHITECTURE}")
-        endif()
+        set(GLIBC_SYSTEM_ROOT "${CMAKE_SYSROOT}")
```
`${SWIFT_SDK_${sdk}_PATH}` ?

gribozavr commented:
LGTM modulo the two comments above.

gribozavr commented:
@swift-ci Please test

karwa commented:
- changed sysroot to use SDK path
- changed cross-compiling check to query if the SDK path is "/". That should be enough to identify the build machine's SDK for all SDKs which generate a glibc modulemap.

rintaro commented on stdlib/public/Platform/CMakeLists.txt:
```diff
@@ -68,11 +63,28 @@ foreach(sdk ${SWIFT_SDKS})
+
+      # If compiling for a non-native SDK, generate a native modulemap without a sysroot prefix
+      # The prefixed one is still the 'glibc.modulemap' which we'll look at during the build process
+      if(NOT "${SWIFT_SDK_${sdk}_PATH}" STREQUAL "/")
```
Sorry, I think, this doesn't work as expected.
As for Android SDK case, the module map should contain
`${SWIFT_SDK_${sdk}_PATH}` prefixed header path even in installed one.

Could you consider to use `if("${sdk}" STREQUAL "${SWIFT_HOST_VARIANT_SDK}" AND "${arch}" STREQUAL "${SWIFT_HOST_VARIANT_ARCH}")`?

gribozavr replied:
If it comes to being that subtle, I'd prefer a hardcoded list so that it is obvious which platform is taking which branch.

karwa replied:
@rintaro I suppose we want to generate a non-prefixed modulemap when:
- We're building the stdlib for the host, and
- The host is not native to the build machine

Would that be sufficient for Android's situation?

rintaro replied:
@karwa I think so.
Or just hardcode to exclude `"ANDROID"` as @gribozavr said would be OK for now.

rintaro commented:
@karwa Could you include `FIXME` in #2473 in this PR?
Then, I can close conflicting my PR.

karwa commented on stdlib/public/Platform/CMakeLists.txt:
```diff
@@ -68,11 +64,34 @@ foreach(sdk ${SWIFT_SDKS})
+      # FIXME: When SDK is a cross-compile target (SDK != Host), the generated
+      #        modulemap will be relative to the Host, with hardcoded paths.
+      #        It is not relocatable to the target platform itself.
+      #        This only affects ANDROID right now, but could affect cross-compiled LINUX targets
```
@rintaro done

gribozavr commented:
I would like to look at the diff again tomorrow, but let's see what CI says.

@swift-ci Please test

gribozavr commented on stdlib/public/Platform/CMakeLists.txt:
```diff
@@ -68,11 +64,34 @@ foreach(sdk ${SWIFT_SDKS})
+      #        This only affects ANDROID right now, but could affect cross-compiled LINUX targets
+        
       swift_install_in_component(sdk-overlay
           FILES "${glibc_modulemap_out}"
```
Shouldn't this install the "native" one modulemap?  How would the compiler know to look inside the `native-substitutes` dir?

karwa replied:
We set `glibc_modulemap_out` to `glibc_sysroot_relative_modulemap_out` (line 84), so it'll install the sysroot-relative one instead. The prefixed one then just becomes a build artefact.

gribozavr replied:
I see now.  Thanks for explaining it to me.

gribozavr commented on stdlib/public/Platform/CMakeLists.txt:
```diff
@@ -37,29 +37,25 @@ foreach(sdk ${SWIFT_SDKS})
-        endif()
+      
+      # Determine the location of glibc headers based on the target.
+      set(GLIBC_NATIVE_INCLUDE_PATH "/usr/include")
```
Could you rename `NATIVE` => `SYSROOT_RELATIVE`?  I think that describes it better.

gribozavr commented on stdlib/public/Platform/CMakeLists.txt:
```diff
@@ -68,11 +64,34 @@ foreach(sdk ${SWIFT_SDKS})
+      # without a sysroot prefix.
+      if(NOT "${SWIFT_SDK_${SWIFT_HOST_VARIANT_SDK}_PATH}" STREQUAL "/")
+
+        set(glibc_native_modulemap_out "${module_dir}/native-substitutes/glibc.modulemap")
```
Could you rename `native-substitutes` => `sysroot-relative-modulemaps`?

gribozavr commented:
@karwa The change LGTM, but it looks like an NFC to me -- I don't understand how the compiler would find the "native" module maps.

springsup committed:
Specify '-resource-dir' when compiling Swift libraries, so we can find
libraries for the destination host when cross-compiling.

karwa commented:
It works because we create two module maps:
- lib/swift/linux/arch/glibc.modulemap: as today. This is what we use when building the standard library (on the build machine), so its headers are prefixed with the sysroot location
- lib/swift/linux/arch/sysroot-relative-modulemaps/glibc.modulemap: which is new, and is what the resulting host should use.

When cross-compiling, the first one is only useful for building the standard library for the host on the build machine. The host can't actually use it, so we only keep it for building purposes and the second one is what we install. I've clarified this in the comments.

Additionally, I have a change which I was going to propose later, but it might make sense to package it with this together: we need to explicitly set the `resource-dir` build flag when compiling swift files. Otherwise the build machine's swift compiler will infer that it is to use its _own_ library directory, which isn't correct when cross-compiling. It means that when building on OSX it won't find either modulemap at all.

gribozavr commented:
@swift-ci Please test

gribozavr commented:
The failures are unrelated.
